### ISSUE TITLE
Same patched return value object returned in parameterized tests

### ISSUE THREAD

**Issue by LKajan** [pytest-dev/pytest#7113]
When mocking a return value of a function the same object is apparently returned in all parametrized test cases. Is this expected behavior? This might be more an issue of a unittest mock library but I'm thinking if something can be made how parametrized test cases are generated and isolated on pytest side.

Here's a small test example of the situation. Actual test doesn't make any sense.

```python
#test_foo.py

from unittest.mock import patch
import pytest

def foo(value):
    return {}

@pytest.mark.parametrize('value', ['a', 'b'])
@patch('test_foo.foo', return_value={'key1': 'value', 'key2': 'value'})
def test_return_value(foo_mock, value):
    a = foo(value)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}

```

```
================================== test session starts ==================================
platform win32 -- Python 3.8.1, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- c:\users\lakaj\projektit\temp\pytest_t\env\scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\LAKAJ\projektit\temp\pytest_t
collected 2 items                                                                                                                                                                          

test_foo.py::test_return_value[a] a<1758705065344>: {'key1': 'value', 'key2': 'value'}
PASSED
test_foo.py::test_return_value[b] a<1758705065344>: {'key1': 'value'}
FAILED

====================================== FAILURES ======================================
__________________________________ test_return_value[b] __________________________________

foo_mock = <MagicMock name='foo' id='1758707357872'>, value = 'b'

    @pytest.mark.parametrize('value', ['a', 'b'])
    @patch('test_foo.foo', return_value={'key1': 'value', 'key2': 'value'})
    def test_return_value(foo_mock, value):
        a = foo(value)
        print(f'a<{id(a)}>: {a}')
>       assert a == {'key1': 'value', 'key2': 'value'}
E       AssertionError: assert {'key1': 'value'} == {'key1': 'value', 'key2': 'value'}
E         Common items:
E         {'key1': 'value'}
E         Right contains 1 more item:
E         {'key2': 'value'}
E         Full diff:
E         - {'key1': 'value', 'key2': 'value'}
E         + {'key1': 'value'}

test_foo.py:16: AssertionError
============================ short test summary info ============================
FAILED test_foo.py::test_return_value[b] - AssertionError: assert {'key1': 'value'} == {'key1': 'value', 'key2': 'value'}
=========================== 1 failed, 1 passed in 0.15s ===========================
```

**environment:**
- Windows 10
- Python 3.8.1
- pytest-5.4.1

**Comment by bluetech** [comment]
The dict from `return_value={'key1': 'value', 'key2': 'value'}` is only created once. If you modify it, it will remain modified. Neither `mock` nor `pytest` will copy it, and it's not very reasonable to do so.

My suggestion is to rewrite the code to use `patch` as a context manager instead of a function decorator.

**Comment by LKajan** [comment]
I understand that the object is created only once and mocked function returns the same object but I would expect the returned object to be the same only inside one test case. I would expect that the two parameterized test cases would have different scope and the mock would be stopped between tests.

Your suggestion to define the return value inside the test function works and that is how I also solved this. I just see that it would be more consistent that decoration way would always work and I don't have to sometimes rely to some other mocking implementation.

**Comment by bluetech** [comment]
Hi @LKajan, can you explain more what you mean by scopes?

Let me rephrase the code in this (equivalent) way:

```py
from unittest.mock import patch
import pytest

def foo(value):
    return {}

RETURN_VALUE = {'key1': 'value', 'key2': 'value'}

@pytest.mark.parametrize('value', ['a', 'b'])
@patch('test_foo.foo', return_value=RETURN_VALUE)
def test_return_value(foo_mock, value):
    a = foo(value)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}
```

I think it makes clear that `RETURN_VALUE` is a single global object that is only created once, and no copies are made of it, so if it is modified in-place somewhere (the `del a['key2']
` in this case), it will remain modified subsequently.

**Comment by LKajan** [comment]
Thanks for getting back to this and thanks for the explanation with that code.

With the scope I meant that now it feels like all the statements in the two test cases are run after each other in one function call with the same scope.
Like:
```python
@patch('test_foo.foo', return_value={'key1': 'value', 'key2': 'value'})
def test_return_value(foo_mock):
    value='a'
    a = foo(value)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}

    value='b'
    a = foo(value)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}
```
And I was expecting it to more like (separed scopes):
```python
@pytest.fixture
def value_a():
    return 'a'

@pytest.fixture
def value_b():
    return 'b'

@patch('test_foo.foo', return_value={'key1': 'value', 'key2': 'value'})
def test_return_value_a(foo_mock, value_a):
    a = foo(value_a)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}

@patch('test_foo.foo', return_value={'key1': 'value', 'key2': 'value'})
def test_return_value_b(foo_mock, value_b):
    a = foo(value_b)
    print(f'a<{id(a)}>: {a}')
    assert a == {'key1': 'value', 'key2': 'value'}
    del a['key2']
    assert a == {'key1': 'value'}
```

I kind of was expecting / hoping that the patch decorator was called again for each test case but now that I have thought this for a while and remembering that decorators are called bottom up it is quite clear that the patch decorator is called only once and the return value is therefore the same object. Apparently swapping the order of patch and parameterize decorators doesn't work and wouldn't know if that would be any reasonable.
